spark3/src/main/java/org/apache/iceberg/actions/Spark3SnapshotAction.java

```diff
@@ -0,0 +1,148 @@
+          taeException);
+    } catch (NoSuchNamespaceException nsnException) {
+      throw new IllegalArgumentException("Cannot create snapshot because the namespace given does not exist",
+          nsnException);
+    }
+
+    boolean threw = true;
+    try {
+      String stagingLocation = icebergTable.location() + "/" + ICEBERG_METADATA_FOLDER;
+      LOG.info("Beginning snapshot of {} to {} using metadata location {}", sourceTableName(), destTableName(),
+          stagingLocation);
+
+      TableIdentifier v1TableIdentifier = v1SourceTable().identifier();
+      SparkTableUtil.importSparkTable(spark(), v1TableIdentifier, icebergTable, stagingLocation);
+      stagedTable.commitStagedChanges();
+      threw = false;
+    } finally {
+      if (threw) {
+        LOG.error("Error when attempting to commit snapshot changes, rolling back");
+        if (stagedTable != null) {
+          stagedTable.abortStagedChanges();
+        }
+      }
+    }
+
+    long numMigratedFiles;
+    Snapshot snapshot = icebergTable.currentSnapshot();
+    numMigratedFiles = Long.valueOf(snapshot.summary().get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
+    LOG.info("Successfully loaded Iceberg metadata for {} files", numMigratedFiles);
+    return numMigratedFiles;
+  }
+
+  private Map<String, String> buildPropertyMap() {
+    Map<String, String> properties = new HashMap<>();
+    properties.put(TableCatalog.PROP_PROVIDER, "iceberg");
+    properties.put(TableProperties.GC_ENABLED, "false");
+    properties.putAll(JavaConverters.mapAsJavaMapConverter(v1SourceTable().properties()).asJava());
+    properties.putAll(additionalProperties());
+
+    // Don't use the default location for the destination table if an alternate has be set
+    if (destTableLocation != null) {
+      properties.putAll(tableLocationProperties(destTableLocation));
+    }
+
+    return properties;
+  }
+
+  @Override
+  protected TableCatalog checkSourceCatalog(CatalogPlugin catalog) {
+    // Currently the Import code relies on being able to look up the table in the session code
+    if (!(catalog.name().equals("spark_catalog"))) {
+      throw new IllegalArgumentException(String.format(
+          "Cannot snapshot a table that isn't in spark_catalog, the session catalog. " +
+              "Found source catalog %s", catalog.name()));
+    }
+
+    if (!(catalog instanceof TableCatalog)) {
+      throw new IllegalArgumentException(String.format(
+          "Cannot snapshot a table from a non-table catalog %s. Catalog has class of %s.", catalog.name(),
+          catalog.getClass().toString()
+      ));
+    }
+
+    return (TableCatalog) catalog;
+  }
+
+  @Override
+  public SnapshotAction withLocation(String location) {
+    Preconditions.checkArgument(!sourceTableLocation().equals(location),
+        "Cannot create snapshot where destination location is the same as the source location. This" +
```

Comment:
I think we miss ` ` at the end of this line. I'd also split it in between sentences.